File: code/PubSub/src/main/scala/PubMW.scala
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
import com.rabbitmq.client.{AMQP, Connection, ConnectionFactory, Channel, MessageProperties}

object JsonSerializationUtil {
  private val mapper = new ObjectMapper() with ScalaObjectMapper
  mapper.registerModule(DefaultScalaModule)

  def serializeToJsonBytes(obj: Any): Array[Byte] = {
    mapper.writeValueAsBytes(obj)
  }
}

object PubMW {
  private var connection: Connection = _
  private var channel: Channel = _
  private var currentTopic: String = _

  def connect_to_server(host: String, port: Int): Unit = {
    val factory = new ConnectionFactory()
    factory.setHost(host)
    factory.setPort(port)
    connection = factory.newConnection()
    channel = connection.createChannel()
  }

  def setup_message_exchange(exchangeName: String): Unit = {
    channel.exchangeDeclare(exchangeName, "fanout", true)
    currentTopic = exchangeName
  }

  def publish_message(message: String): Unit = {
    val serializedMessage = JsonSerializationUtil.serializeToJsonBytes(message)
    val properties = MessageProperties.PERSISTENT_TEXT_PLAIN

    if (currentTopic.isEmpty) {
      throw new IllegalStateException("Exchange not initialized")
    }

    channel.basicPublish(currentTopic, "", properties, serializedMessage)
  }

  def close_connection(): Unit = {
    if (channel != null) channel.close()
    if (connection != null) connection.close()
  }
}

File: PubSubRaft/src/main/kotlin/PubMW.kt
package org.example

import com.rabbitmq.client.Connection
import com.rabbitmq.client.ConnectionFactory
import com.rabbitmq.client.Channel
import com.rabbitmq.client.MessageProperties
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule

// Define an interface for handling upcall events
interface UpcallHandler {
    fun onMessagePublished(topic: String, message: String)
    fun onErrorOccurred(error: Throwable)
}

object PubMW {
    private var connection: Connection? = null
    var channel: Channel? = null
    var currentTopic: String? = null
    private var upcallHandler: UpcallHandler? = null

    // Initialize ObjectMapper for JSON serialization
    var mapper: ObjectMapper = jacksonObjectMapper().apply {
        registerKotlinModule()
    }

    fun connectToServer(host: String, port: Int) {
        val factory = ConnectionFactory()
        factory.host = host
        factory.port = port
        connection = factory.newConnection()
        channel = connection?.createChannel()
    }

    fun setupMessageExchange(exchangeName: String) {
        channel?.exchangeDeclare(exchangeName, "fanout", true)
        currentTopic = exchangeName
    }

    fun publishMessage(message: String) {
        try {
            val serializedMessage = serializeToJsonBytes(message)
            val properties = MessageProperties.PERSISTENT_TEXT_PLAIN

            currentTopic?.let {
                channel?.basicPublish(it, "", properties, serializedMessage)
                upcallHandler?.onMessagePublished(it, message)
            } ?: throw IllegalStateException("Exchange not initialized")
        } catch (e: Throwable) {
            upcallHandler?.onErrorOccurred(e)
        }
    }

    fun setUpcallHandler(handler: UpcallHandler) {
        upcallHandler = handler
    }

    fun closeConnection() {
        channel?.close()
        connection?.close()
    }

    fun serializeToJsonBytes(obj: Any): ByteArray {
        return mapper.writeValueAsBytes(obj)
    }
}
